### PR TITLE
Factor analysis

### PR DESCRIPTION
**Pull request by sophieball**
#59 

The R code plots eigenvalues and tries different numbers of factors.

### CODE REVIEW

**Comment by bvasiles** [comment]
Can I see the output plots?

File: main/factor.R [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# redirect output to a file
+sink("factor.out")
+
+library(readr)
+library(nFactors)
```
Comment (CaptainEmerson) [inline main/factor.R:5]: how important is this package? it's not easily available to me. it's also not in the CSV below?

File: r_packages.csv [hunk 2]
```diff
@@ -1,5 +1,6 @@
 name,version,sha256
 BH,1.72.0-3,888ec1a3316bb69e1ba749b08ba7e0903ebc4742e3a185de8d148c13cddac8ab
+DataCombine,0.2.21,352b235612e2cf8234b3ab5f9aa6f7a394b006b98d24e315940ccc65c4218b47
```
Comment (CaptainEmerson) [inline r_packages.csv:3]: no readr in this file?

File: main/factor.R [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ev <- eigen(cor(dat, method="spearman", use="complete.obs"))
+write(paste("Eigen values:", ev$values), stdout())
+
+ap <- parallel(subject=nrow(dat), var=ncol(dat), rep=100, cent=0.05)
+nS <- nScree(x=ev$values, aparallel=ap$eigen$qevpea)
+png("scree.png")
+plotnScree(nS)
+
+# try different numbers of factors
+for (n_fac in 3:6){
```
Comment (CaptainEmerson) [inline main/factor.R:25]: what does this range mean?

File: main/factor.R [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# redirect output to a file
+sink("factor.out")
+
+library(readr)
+library(nFactors)
+library(DataCombine)
```
Comment (CaptainEmerson) [inline main/factor.R:6]: Hmm, I don't have access to this library either. Hmm...

File: main/factor.R (hunk 1 above)
Comment (CaptainEmerson) [reply main/factor.R:5]: I *think* I can work around these by just using the bazel binary, which should pull in the external packages from CRAN.

File: r_packages.csv (hunk 2 above)
Comment (CaptainEmerson) [reply r_packages.csv:3]:
```
myproject) emersonm@emersonm:~/toxicity-detector$ bazel run main:factor
DEBUG: Rule 'com_grail_rules_r' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "705c63ffc8e7ab4c9626d2be658d6b55e803d9133e8ef0384d1049b13b8bfd99"
DEBUG: Repository com_grail_rules_r instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule http_archive defined at:
  /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
INFO: Analyzed target //main:factor (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //main:factor up-to-date:
  bazel-bin/main/factor
INFO: Elapsed time: 0.366s, Critical Path: 0.05s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
Error in library(readr) : there is no package called ‘readr’
Execution halted
(myproject) emersonm@emersonm:~/toxicity-detector$ 

```

**Comment by CaptainEmerson** [comment]
One more comment: please add datetime to filename so as not to overwrite.

File: main/factor.R (hunk 3 above)
Comment (sophieball) [reply main/factor.R:25]: It tries out different numbers of factors - 20 columns down to 3/4/5/6 factors. Then we can look at the output and see which number makes the most sense.

File: main/factor.R (hunk 1 above)
Comment (sophieball) [reply main/factor.R:5]: hmmm. If it's about `readr`, is in all R files I submitted. I use it to read in csv. It's also in the `r_packages.csv`
If it's about `nFactors`, I'm adding it to the csv

File: r_packages.csv (hunk 2 above)
Comment (sophieball) [reply r_packages.csv:3]: Hmmmm... I can see `readr` in line 53 in this file.

Comment (sophieball) [reply r_packages.csv:3]: Hmmmm.. I can see `readr` in line 53

File: main/factor.R (hunk 4 above)
Comment (sophieball) [reply main/factor.R:6]: Adding it to `r_packages.csv`

Comment (sophieball) [reply main/factor.R:6]: It's in `r_pacakges.csv`. maybe when you run it as a bazel binary you'll be able to use the library?

Comment (CaptainEmerson) [reply main/factor.R:6]: Yes, that's my hope.

File: r_packages.csv (hunk 2 above)
Comment (CaptainEmerson) [reply r_packages.csv:3]: So it is. Still, I blaze ran again and got the same error message. I don't see readr in the built libraries:

```
(myproject) emersonm@emersonm:~/toxicity-detector/bazel-bin/main/factor.runfiles$ ls
__main__  MANIFEST  R_assertthat  R_cli  R_crayon  R_DataCombine  R_data_table  R_digest  R_dplyr  R_ellipsis  R_fansi  R_generics  R_glue  R_lifecycle  R_magrittr  R_pillar  R_pkgconfig  R_purrr  R_R6  R_rlang  R_tibble  R_tidyselect  R_utf8  R_vctrs
```

Does it work (at least get through the imports) when you run it through "blaze run"?  If so, what built libraries do you get?

Comment (sophieball) [reply r_packages.csv:3]: OH! I realized the problem - I've been running it through `feed_data` binary, which has all the dependencies declared in the `BUILD` file. Sorry about that~ I've pushed the updated `BUILD` file

**Comment by CaptainEmerson** [comment]
Looks like travis is failing. I also get a build failure:

```
myproject) emersonm@emersonm:~/toxicity-detector$ bazel build main:factor
INFO: Repository R_nFactors instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule r_repository defined at:
  /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/com_grail_rules_r/R/repositories.bzl:78:31: in <toplevel>
ERROR: An error occurred during the fetch of repository 'R_nFactors':
   No sources found for repository '@R_nFactors'. Perhaps this package is not available for your R version.
INFO: Repository R_psych instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule r_repository defined at:
  /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/com_grail_rules_r/R/repositories.bzl:78:31: in <toplevel>
ERROR: /usr/local/google/home/emersonm/toxicity-detector/main/BUILD:56:9: //main:factor depends on @R_nFactors//:R_nFactors in repository @R_nFactors which failed to fetch. no such package '@R_nFactors//': No sources found for repository '@R_nFactors'. Perhaps this package is not available for your R version.
ERROR: Analysis of target '//main:factor' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.874s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
    Fetching @R_readr; fetching
    Fetching /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/R_readr; Extracting /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/R_readr/readr_1.3.1.tar.gz
(myproject) emersonm@emersonm:~/toxicity-detector$ 


```

**Comment by sophieball** [comment]
> Looks like travis is failing. I also get a build failure:
> 
> ```
> myproject) emersonm@emersonm:~/toxicity-detector$ bazel build main:factor
> INFO: Repository R_nFactors instantiated at:
>   no stack (--record_rule_instantiation_callstack not enabled)
> Repository rule r_repository defined at:
>   /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/com_grail_rules_r/R/repositories.bzl:78:31: in <toplevel>
> ERROR: An error occurred during the fetch of repository 'R_nFactors':
>    No sources found for repository '@R_nFactors'. Perhaps this package is not available for your R version.
> INFO: Repository R_psych instantiated at:
>   no stack (--record_rule_instantiation_callstack not enabled)
> Repository rule r_repository defined at:
>   /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/com_grail_rules_r/R/repositories.bzl:78:31: in <toplevel>
> ERROR: /usr/local/google/home/emersonm/toxicity-detector/main/BUILD:56:9: //main:factor depends on @R_nFactors//:R_nFactors in repository @R_nFactors which failed to fetch. no such package '@R_nFactors//': No sources found for repository '@R_nFactors'. Perhaps this package is not available for your R version.
> ERROR: Analysis of target '//main:factor' failed; build aborted: Analysis failed
> INFO: Elapsed time: 0.874s
> INFO: 0 processes.
> FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
>     Fetching @R_readr; fetching
>     Fetching /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/R_readr; Extracting /usr/local/google/home/emersonm/.cache/bazel/_bazel_emersonm/944306df6cbfee92be3237efb3eb9146/external/R_readr/readr_1.3.1.tar.gz
> (myproject) emersonm@emersonm:~/toxicity-detector$ 
> ```

What's your R version?

**Comment by sophieball** [comment]
The `nFactors` library is needed for eigenvalue analysis. It requires R version >= 3.5.0 (travis uses 3.4.4). 
According to R bazel's [doc](https://github.com/grailbio/rules_r), 

> These rules assume that you have R installed on your system (we recommend 3.4.3 or above), and can be located using the PATH environment variable.
NOTE: After re-installing or upgrading R, please reset the registered toolchain with bazel sync --configure to rebuild your packages with the new installation.

So in my `.bashrc`, I have `export PATH=~/R-4.0.2/bin:$PATH`

Since now you have your own version of R, you might be able to install the packages needed by `factor.R` and run it from your feed data R script using something like `system2("factor.R", xxxxxx)`?

What's the reason that G uses an old version of R?

**Comment by CaptainEmerson** [comment]
> Since now you have your own version of R, you might be able to install the packages needed by factor.R and run it from your feed data R script using something like system2("factor.R", xxxxxx)?

Yea, I just thought of that, too. That's easier than changing the environment. I'll let you know how it goes.

> What's the reason that G uses an old version of R?

IDK why we use the version we use.

**Comment by CaptainEmerson** [comment]
Update: I tried with R 3.6, and nFactors wouldn't install properly. Trying 4.0.2.

**Comment by CaptainEmerson** [comment]
Welp, I tried calling R 4.0.2 from R 3.x, through system2 with no luck. Sophie, could you check to see what libraries Lan usually uses?

**Comment by sophieball** [comment]
> Welp, I tried calling R 4.0.2 from R 3.x, through system2 with no luck. Sophie, could you check to see what libraries Lan usually uses?

I left Lan a message earlier.


I just succeeded in making travis install R4.0 and the build passed! Now you can use the bazel `factor` r binary~!~~~!~~

**Comment by sophieball** [comment]
I got code from Lan. I can push a library-free version later.

**Comment by sophieball** [comment]
@CaptainEmerson I've pushed the library-free version. It produces the same output.

**Comment by CaptainEmerson** [comment]
```
Error in solve.default(cv) : 
  Lapack routine dgesv: system is exactly singular: U[33,33] = 0
Calls: factanal -> diag -> solve -> solve.default
Execution halted

```

**Comment by sophieball** [comment]
> ```
> Error in solve.default(cv) : 
>   Lapack routine dgesv: system is exactly singular: U[33,33] = 0
> Calls: factanal -> diag -> solve -> solve.default
> Execution halted
> ```

Is 33 the number of columns? 

I know singular implies that the data has rows or columns that are all 0s, or some columns are linear combination of others...

**Comment by sophieball** [comment]
@CaptainEmerson I've added some checks to `factor_lan.r` - it will check if there's any cols with no variance and if any cols are linear combinations of the others - then remove those cols.. Can you run it and see if it works?

**Comment by CaptainEmerson** [comment]
```
Warning messages:
1: In FUN(newX[, i], ...) : NAs introduced by coercion
2: In FUN(newX[, i], ...) : NAs introduced by coercion
3: In FUN(newX[, i], ...) : NAs introduced by coercion
Error in qr.default(object) : NA/NaN/Inf in foreign function call (arg 1)
Calls: findLinearCombos ... internalEnumLC -> qr.R -> is.qr -> qr -> qr.default
In addition: Warning message:
In storage.mode(x) <- "double" : NAs introduced by coercion
Execution halted

```

**Comment by CaptainEmerson** [comment]
[1] "dataframe shape: 2249" "dataframe shape: 4"   
[1] "columns with no variance: "
[1] "dataframe shape after droping no var cols: 2249"
[2] "dataframe shape after droping no var cols: 4"

**Comment by sophieball** [comment]
> [1] "dataframe shape: 2249" "dataframe shape: 4"
> [1] "columns with no variance: "
> [1] "dataframe shape after droping no var cols: 2249"
> [2] "dataframe shape after droping no var cols: 4"

dataframe shapes should contain 2 numbers - (num rows, num cols)

**Comment by CaptainEmerson** [comment]
```
Sorry, that was a red herring. Here's the right run:

[1] "dataframe shape: 210" "dataframe shape: 33" 
[1] "columns with no variance: "
[1] "dataframe shape after droping no var cols: 210"
[2] "dataframe shape after droping no var cols: 33" 
$linearCombos
$linearCombos[[1]]
[1] 26  3


$remove
[1] 26

[1] "dataframe shape after dropping linear combination cols: 210"
[2] "dataframe shape after dropping linear combination cols: 33" 

Error in solve.default(cv) : 
  Lapack routine dgesv: system is exactly singular: U[33,33] = 0
Calls: factanal -> diag -> solve -> solve.default
Execution halted

```

**Comment by sophieball** [comment]
> [1] "dataframe shape after dropping linear combination cols: 210"
> [2] "dataframe shape after dropping linear combination cols: 33"

Oh! Sorry - linear combination check returns column numbers not names. Pushed new code..
The message here at least told us there is linear dependency between columns - hence singular matrix

I also made the output more readable, I think you should expect something like:

```
[1] "initial dataframe shape:"
[1] 210 33
[1] "columns with no variance:"
numeric(0)
[1] "dataframe shape after droping no var cols: 210"
[2] "dataframe shape after droping no var cols: 33"
$linearCombos
$linearCombos[[1]]
[1] 26 3


$remove
[1] 26

[1] "dataframe shape after dropping linear combination cols:"
[1] 210 32
```

File: main/factor_lan.R [hunk 5]
```diff
@@ -0,0 +1,72 @@
+# Code by Lan Cheng
+library(caret)
+
+# redirect output to a file
+sink("factor.out")
+
+# Factor analysis to pushback metrics to 3-7 factors
+dat <- read.csv("/dev/stdin")
+dat$comb <- 2*dat$mot1 + 3*dat$mot2
```
Comment (CaptainEmerson) [inline main/factor_lan.R:9]: this shouldn't be here, right?

File: main/factor_lan.R [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+no_var <- as.numeric(which(apply(dat, 2, var) == 0))
+print(paste("columns with no variance:", no_var))
+# Remove columns with 0 variance
+if (length(no_var) > 0){
+  dat <- dat[ - no_var]
+}
+print(paste("dataframe shape after droping no var cols:", dim(dat)))
+
+linear_combos <- findLinearCombos(dat)
+linear_combos
```
[the remark this answers is not in the document]
Comment (sophieball) [reply main/factor_lan.R]: But that tells you which cols have dependencies. I'd rather keep it and add a sentence in the output about what it is.

File: main/factor_lan.R (hunk 5 above)
Comment (sophieball) [reply main/factor_lan.R]: Right!!!! That was my test!!! I'll remove it =p

**Pull request merged**